### PR TITLE
[Bugfix] Provile bitStreamArgs for HLS fMP4

### PR DESCRIPTION
**Pull request by eyezak**
Fix audio sync in HLS fMP4

**Changes**
Provide audio bitstream arguments to ffmpeg for both ts and fMP4 HLS streaming

### CODE REVIEW

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 1]
```diff
@@ -1676,8 +1677,8 @@ public bool CanStreamCopyVideo(EncodingJobInfo state, MediaStream videoStream)
             }
 
             // Source and target codecs must match
-            if (string.IsNullOrEmpty(videoStream.Codec)
-                || !state.SupportedVideoCodecs.Contains(videoStream.Codec, StringComparison.OrdinalIgnoreCase))
+            if (string.IsNullOrEmpty(videoStream.Codec) || (state.SupportedVideoCodecs.Any()
+                && !state.SupportedVideoCodecs.Contains(videoStream.Codec, StringComparison.OrdinalIgnoreCase)))
```
Comment (crobibero) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:1681]:
```suggestion
            if (string.IsNullOrEmpty(videoStream.Codec)
                || (state.SupportedVideoCodecs.Length != 0
                    && !state.SupportedVideoCodecs.Contains(videoStream.Codec, StringComparison.OrdinalIgnoreCase)))
```

**Review by crobibero — APPROVED** [review]

**Comment by nyanmisaka** [comment]
Does `-bsf:a aac_adtstoasc` work on source video containers other than MPEG-TS?

**Comment by eyezak** [comment]
> Does `-bsf:a aac_adtstoasc` work on source video containers other than MPEG-TS?

It definitely applies to raw aac files as well. I can rework this change to look at source files instead of output, if that is correct here

**Comment by nyanmisaka** [comment]
> It definitely applies to raw aac files as well. I can rework this change to look at source files instead of output, if that is correct here

That make sense. It should be applied when mediaSourceContainer is aac.

**Comment by Shadowghost** [comment]
What's the state on this?

**Review by joshuaboniface — APPROVED** [review]

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 2]
```diff
@@ -1680,7 +1680,7 @@ private string GetAudioArguments(StreamState state)
             {
                 if (EncodingHelper.IsCopyCodec(audioCodec))
                 {
-                    var bitStreamArgs = EncodingHelper.GetAudioBitStreamArguments(state, state.Request.SegmentContainer, state.MediaSource.Container);
+                    var bitStreamArgs = EncodingHelper.GetAudioBitStreamArguments(state, state.Request.SegmentContainer, state.MediaSource.Container, state.TranscodingType);
```
Comment (nyanmisaka) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:1683]:
```suggestion
                    var bitStreamArgs = EncodingHelper.GetAudioBitStreamArguments(state, state.Request.SegmentContainer, state.MediaSource.Container);
```

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 3]
```diff
@@ -1711,7 +1711,7 @@ private string GetAudioArguments(StreamState state)
             if (EncodingHelper.IsCopyCodec(audioCodec))
             {
                 var videoCodec = _encodingHelper.GetVideoEncoder(state, _encodingOptions);
-                var bitStreamArgs = EncodingHelper.GetAudioBitStreamArguments(state, state.Request.SegmentContainer, state.MediaSource.Container);
+                var bitStreamArgs = EncodingHelper.GetAudioBitStreamArguments(state, state.Request.SegmentContainer, state.MediaSource.Container, state.TranscodingType);
```
Comment (nyanmisaka) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:1714]:
```suggestion
                var bitStreamArgs = EncodingHelper.GetAudioBitStreamArguments(state, state.Request.SegmentContainer, state.MediaSource.Container);
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 4]
```diff
@@ -950,14 +950,15 @@ public static string GetBitStreamArgs(MediaStream stream)
             }
         }
 
-        public static string GetAudioBitStreamArguments(EncodingJobInfo state, string segmentContainer, string mediaSourceContainer)
+        public static string GetAudioBitStreamArguments(EncodingJobInfo state, string segmentContainer, string mediaSourceContainer, TranscodingJobType transcodingType)
```
Comment (nyanmisaka) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:953]:
```suggestion
        public static string GetAudioBitStreamArguments(EncodingJobInfo state, string segmentContainer, string mediaSourceContainer)
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 5]
```diff
@@ -950,14 +950,15 @@ public static string GetBitStreamArgs(MediaStream stream)
...
         {
             var bitStreamArgs = string.Empty;
             var segmentFormat = GetSegmentFileExtension(segmentContainer).TrimStart('.');
 
             // Apply aac_adtstoasc bitstream filter when media source is in mpegts.
             if (string.Equals(segmentFormat, "mp4", StringComparison.OrdinalIgnoreCase)
                 && (string.Equals(mediaSourceContainer, "mpegts", StringComparison.OrdinalIgnoreCase)
+                    || transcodingType == TranscodingJobType.Hls
```
Comment (nyanmisaka) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:961]:
```suggestion
                    || string.Equals(mediaSourceContainer, "aac", StringComparison.OrdinalIgnoreCase)
```

**Review by nyanmisaka — CHANGES_REQUESTED** [review]

**Review by nyanmisaka — APPROVED** [review]

**Pull request merged**